docs/DesignDocs/SGX_QuoteVerify_Integration.md

```diff
@@ -0,0 +1,522 @@
+  - Call QVL API to verify the quote
+
+#### Add OCALL to support enclave-side SGX ECDSA-p256 quote verification
+As quote verification will be done by DCAP QVL/QvE, for enclave-side plugin,
+it requires OCALL to call host-side APIs.
+
+In this proposal, we will add one OCALL `oe_verify_quote_ocall()` in
+`edl/sgx/attestation.edl`.
+The OCALL is used for passing quote buffer and related parameters to host side.
+All other relevant logic will be implemented on the host side, in files
+`host/sgx/ocalls.c` and `host/sgx/quote.c`.
+If DCAP QVL cannot be found in host-side plugin, a specific error will be returned.
+
+Note that currently only ECDSA-p256 quotes are supported, but the OCALL will keep
+`format_id` and `opt_params` for forward compatibility.
+
+```C
+oe_result_t oe_verify_quote_ocall(
+    [in] const oe_uuid_t* format_id,
+    [in, size=opt_params_size] const void* opt_params,
+    size_t opt_params_size,
+    [in, size=quote_size] const void* p_quote,
+    size_t quote_size,
+    oe_datetime_t expiration_check_date,
+    [out] uint32_t *p_collateral_expiration_status,
+    [out] uint32_t *p_quote_verification_result,
+    [in, out, size=qve_report_size] void* p_qve_report_info,
+    size_t qve_report_size,
+    [out, size=supplemental_data_size] void* p_supplemental_data,
+    size_t supplemental_data_size,
+    [out] size_t* p_supplemental_data_size_out);
+```
+
+#### Update implementation of existing plugin functions
+These plugin functions include `_verify_evidence()`, `_verify_report()`
+and they both invoke function `oe_verify_quote_with_sgx_endorsements()`
+for quote verification.
+
+- Update API `oe_verify_quote_with_sgx_endorsements()` implementation as below:
+  - Try to call DCAP QVL to verify quote first
+    - In host-side plugin:
+      - Call host-side function `oe_sgx_verify_quote()` with SGX endorsements,
+        which will call SGX QVL library to verify quote.
+    - In enclave-side plugin:
+      - Construct structure `sgx_ql_qe_report_info_t` as below, `nonce` and
+        `app_enclave_target_info` are input, qve_report is QvE report which
+        target to application enclave.
+      ```C
+      typedef struct _sgx_ql_qe_report_info_t {
+        sgx_quote_nonce_t nonce;
+        sgx_target_info_t app_enclave_target_info;
+        sgx_report_t qve_report;
+      }sgx_ql_qe_report_info_t;
+      ```
+      - Call OCALL function `oe_verify_quote_ocall()` to host-side plugin library
+        to verify SGX quote.
+      - With all the QvE return values, call new API.
+        `oe_verify_qve_report_and_identity()` (described below) to
+        verify QvE report and identity
+  - If DCAP QVL returns error to indicate that QVL doesn't not exist on current
+    platform, then automatically fallback to the existing implementation
+  - Add log information to indicate whether the QVL-based or existing implementation
+    is used
+- If QVL-based quote verification is used, in order to exactly align with current
```

Comment:
Current implementation also accepts "SWHardeningNeeded" as a success case. See [tcbinfo.c](https://github.com/openenclave/openenclave/blob/master/common/sgx/tcbinfo.c#L243).

Reply:
Thanks, I will update the design doc to address this issue

Reply:
BTW, the comment in [tcbinfo.c](https://github.com/openenclave/openenclave/blob/master/common/sgx/tcbinfo.c#L243) is not accurate, only on the old platform which will be impacted by LVI, Intel will mark UpToDate tcb as SWHardeningNeeded